### DISCUSSION TITLE
Is it possible to implement some sort of "preview text"

### DISCUSSION THREAD

**Discussion by zaksmolen** [icsharpcode/AvalonEdit#411] [Q&A]
We use AvalonEdit for a built-in editor for our data files, which are text-based but often tabular in nature. We are looking at a feature which will populate some things in the file, visually, but I don't want to just paste the text in. Essentially I want something like the "preview text" that many IDEs have when they auto-suggest things (see screenshot from VS). But this is distinct from the dropdown popup for auto-complete... I want text that is visible in the line, but not selectable or actually there. I'm happy to dive deep to look into implementing this, but a. I wanted to see if this was already available in some capacity and I just didn't know the right keywords to use, or b. if anyone had already looked into something like this before.
![image](https://github.com/icsharpcode/AvalonEdit/assets/12516896/89d377ba-2629-489a-8031-36a4730c0405)

**Comment by goswinr** [comment]
The same way you draw red squiggles for errors you could draw gray text  in a VisualLine, that is not part of the TextDocument. No?

**Reply by zaksmolen** [reply]
That is a fantastically simple approach! I'll take a look into that, thanks!

**Reply by zaksmolen** [reply]
Unfortunately this is very simple in theory but seems like it will be exceedingly difficult in practice. It's looking like I'll need to create a completely new element generator and/or a layer, something much lower level than the decorators, if I want to render custom stuff. Ideally I wanted to hook into the visual line system so I didn't have to pre-calculate every single field in the file and instead we'd just populate the stuff from lookups and metadata as it comes into view. These files can be incredibly large so efficiency is a concern. 

I tried tying into some transformers we had for testing, for example a colorizing transformer that was already checking the metadata I'll be using to color specific columns different colors, but there was no way for me to access the right sort of thing in the VisualLine. I even tried swapping the VisualLineElement with a FormattedTextElement in the parent VisualLineElement but it seems like unless an internal generator is creating these elements the TextRunProperties will be null which causes a variety of other issues...

I then went the route of looking at how the folding manager takes text and turns it into those little folded blocks, which seems more like what I'm looking for, but incredibly involved to implement my own version (and possibly overkill). Essentially I want to be able to take a VisualLine in the file, get the line behind it, use some lookups to get the metadata for that line, and if we have some calculated fields I want to render some text in those spots in the file. So for example:

```
  A  |  B
  2  |  
  3  |  
```
Given this in my file, I want to calculate something off of the value that's in the A column and display that (visually but not change the underlying text) in the editor. So say we want to display value * 2 we'd end up with 

```
  A  |  B
  2  |  4
  3  |  6
```
But the "4" and "6" would be grayed out.

I'll keep poking around though, because in theory what you said is exactly what I want. In the same way that we're rendering the decorators, I want to render text on the screen in specific column positions for specific lines in the files regardless of the actual text that is there. It just seems like a lot of the things I'd need to do it on top of existing mechanisms without creating a renderer from the bottom up are all internally protected.

**Reply by goswinr** [reply]
Did you try to draw with the interface IBackgroundRenderer? I think it gives you a DrawingContext where you can draw anything.

**Reply by zaksmolen** [reply]
I have a somewhat working prototype now actually! I took some insights from here https://stackoverflow.com/questions/28379206/custom-hyperlinks-using-avalonedit on how to build my own `VisualLineElementGenerator` and it's being applied as expected. I'll see about sharing a simplified code example when I work out more of the issues.

**Reply by zaksmolen** [reply]
It is working well for rows that have space padding along the full width of the line, but I'm running into a weird issue that I'm not sure how to resolve.
![image](https://github.com/icsharpcode/AvalonEdit/assets/12516896/6bb5ed75-0382-4d03-99ca-ea66ac1c6f59)
So given this example let's pretend that the "GENERATED" text is what my line creates. The problem I have here is that my virtual text that I'm inserting might overshoot the line boundaries. For example I have one situation specifically where it's crashing when
`offset == CurrentContext.VisualLine.LastDocumentLine.EndOffset` in my `ConstructElement(int offset)`.

It fails in `TextView.BuildVisualLine()` complaining `System.InvalidOperationException: 'Line N was skipped by a VisualLineElementGenerator, but it is not collapsed.'` I assume in this case essentially it's overwriting the newline and bleeding over into the next line, causing it to be skipped? But I don't know what to change or see anything that I think I can change to essentially add this line into the visual here. I had hoped virtual spaces might do something here, but no dice. 

Any ideas on how to handle this situation? To greatly simplify: I have a `VisualLineElementGenerator` that generates possible visual elements that should exist at or after the end of the original line.

Edit:
Just to follow up with more info from my debugging, in `ConstructElement` the rows that are failing seem to be where `CurrentContext.VisualLine` has multiple `DocumentLines` in it, which makes sense why it's complaining about start != end but not collapsed. This table isn't really visually different than the previous two that work though, other than having inconsistent space padding at the end of the rows. Line endings are the same.
